src/server/modemwatcher.cpp

```diff
@@ -31,15 +33,37 @@ ModemWatcher::ModemWatcher(const QString objectPath, const QString interface, QO
     QObject(parent), m_objectPath(objectPath), m_interface(interface), m_interfaceAvailable(false)
 {
     QDBusConnection::systemBus().connect(OfonoConstants::OFONO_SERVICE, m_objectPath,
-                                         OfonoConstants::OFONO_MODEM_INTERFACE, "PropertyChanged",
-                                         this, SLOT(onModemPropertyChanged(QString, QDBusVariant)));
+                                         OfonoConstants::OFONO_MODEM_INTERFACE,
+                                         "PropertyChanged", this,
+                                         SLOT(onModemPropertyChanged(QString, QDBusVariant)));
 
+    m_ofonoWatcher = new QDBusServiceWatcher(OfonoConstants::OFONO_SERVICE,
+                                             QDBusConnection::systemBus(),
+                                             QDBusServiceWatcher::WatchForRegistration,
+                                             this);
+    connect(m_ofonoWatcher, SIGNAL(serviceRegistered(QString)),
+            this, SLOT(getProperties()));
+
+    if (QDBusConnection::systemBus().interface()->isServiceRegistered(OfonoConstants::OFONO_SERVICE))
```

Comment:
Can't deduce from QDBusServiceWatcher docs if it checks initial condition. It would be logical if it would - in which case it might be enough to just wait for signals from it?

And even if it does not, then the availability could be tracked by doing the synchronous query once, caching the result & updating it on serviceRegistered & serviceUnregistered signals -> eliminates need for multiple synchronous queries.